### DISCUSSION TITLE
Wordpress - maybe ray should be registered as a mu-plugin?

### DISCUSSION THREAD

**Discussion by cjke** [spatie/wordpress-ray#31] [Ideas]
At the moment ray registers itself in Wordpress as a standard plugin, this makes the load order dependent on the names of the plugins, alphabetically. 

For example (from wp_options.active_plugins in the db):

```
[
     "my-cool-plugin/my-cool-plugin.php",  👎 
     "spatie-ray/wp-ray.php", 🕶️ 
     "www-projects/my-other-cool-plugin.php", 👍 
 ]
```

So `ray()` will work for `my-other-cool-plugin.php` but not `my-cool-plugin`. This applies to early plugin methods such as `my-cool-plugin@__construct`. By* setting it up as a mu-plugin it _should_ have a higher priority. 

I think this might cause some deceptive bugs if people aren't aware of the order (tbf, I wasn't until I googled around to figure out why i couldn't call ray()).

It might be possible to symlink the ray plugin root file over to mu-plugins (mu-plugins have a different directory structure, so you might not be able to simply move it).

**Comment by freekmurze** [comment]
@jeffreyvr Are you familiar with `mu-plugins`? What would be the cleanest way to go about this?

**Reply by cjke** [reply]
Thanks for reading the discussion. 

Just a disclaimer I'm not an expert at Wordpress architecture, I just occasionally use it for smaller clients. So mu-plugins might not be best practice. 
In the past when I've used kint-php (kind of similar to ray, but not as good as ray) within Wordpress I just added the following to my wp-config.php:

```
define( 'WP_DEBUG', true );

require_once(__DIR__ . '/kint.phar');
Kint\Renderer\RichRenderer::$folder = false;

/* That's all, stop editing! Happy publishing. */
```

Again, not sure if best practice, but it was the only way I could get kint booted before any plugin and before any theme.

**Reply by freekmurze** [reply]
Thanks for creating the discussing. I, like you, am not a WordPress expert. I don't use at all. That's why I'm asking @jeffreyvr for an opinion as he seems quite knowledgable how to get things done in the WordPress world.

**Reply by jeffreyvr** [reply]
Interesting point. I personally haven't yet needed to log something with Ray before the `plugins_loaded` hook is fired, so I haven't run into this issue yet. 

I'm not sure if this would be needed often, but I guess there could be a mention about this in the documentation where moving the plugin to the `mu-plugins` folder ([caveats)](https://wordpress.org/support/article/must-use-plugins/#caveats)) or a creating symlink could be suggested.

**Reply by cjke** [reply]
Documentation note sounds good to me, and an end user like myself using the proxy solution mentioned in the caveat link sounds like a fair compromise (similar to a symlink)

**Reply by freekmurze** [reply]
@cjke could you PR a note to our docs? There's an edit button on top of each page to easily create a PR. 👍

**Reply by cjke** [reply]
Cool can do, but won't be until the weekend (working around the clock atm)

**Reply by freekmurze** [reply]
That's fine, no rush!

**Reply by cjke** [reply]
Looks like maplerock beat me to it, found in commit: https://github.com/chrisk-7777/ray/commit/dbaadcad9fbccac1af3579e200b6de7e0a105481 
Thanks again for the awesome product
